documentation/src/main/asciidoc/mapper-orm-bridge.asciidoc

```diff
@@ -0,0 +1,1353 @@
+and get a reference to to that object.
+<2> Add a value to the sub-fields for the object we just added.
+Note we're calling `addValue` on the object we just added, not on `target`.
+====
+
+=== Object field storage
+
+By default, object fields are flattened,
+meaning that the tree structure is not preserved.
+See <<mapper-orm-indexedembedded-storage-flattened>> for more information.
+
+It is possible to switch to <<mapper-orm-indexedembedded-storage-nested,nested storage>>
+by passing an argument to the `objectField` method, as shown below.
+Each value of the object field will then transparently be indexed as a separate nested document,
+without any change to the `write` method of the bridge.
+
+.Declaring an object field as nested
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/bridge/document/model/dsl/object/InvoiceLineItemsDetailBinder.java[tags=bind]
+----
+<1> Declare an object field with `objectField`.
+<2> Pass the name of the object field.
+<3> Pass the storage type of the object field, here `NESTED`.
+<4> Declare the object field as multi-valued.
+<5> Get a reference to the declared object field
+and pass it to the bridge for later use.
+<6> Create sub-fields, get references to these fields
+and pass them to the bridge for later use.
+====
+
+[[mapper-orm-bridge-index-field-type-dsl]]
+== Defining index field types
+
+=== Basics
+
+A specificity of Lucene-based search engines (including Elasticsearch) is that field types
+are much more complex than just a data type ("string", "integer", ...).
+
+When declaring a field, you must not only declare the data type,
+but also various characteristics that will define how the data is stored exactly:
+is the field sortable,
+is it projectable,
+is it analyzed and if so with which analyzer,
+...
+
+Because of this complexity,
+when field types must be defined in the various binders
+(`ValueBinder`, `PropertyBinder`, `TypeBinder`),
+they are defined using a dedicated DSL.
+
+The entry point to this DSL is the `IndexFieldTypeFactory`.
+The type factory is generally accessible though the context object passed to the binders
+(`context.getTypeFactory()`).
+In the case of `PropertyBinder`, `TypeBinder`,
+the type factory can also be passed to the lambda expression passed to the `field` method
+to define the field type inline.
+
+The type factory exposes various `as*()` methods,
+for example `asString` or `asLocalDate`.
+These are the first steps of the type definition DSL,
+where the data type is defined.
+They return other steps, from which options
+can be set, such as the analyzer.
+See below for an example.
+
+.Defining a field type
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/bridge/types/dsl/simple/ISBNBinder.java[tags=include]
+----
+<1> Get the `IndexFieldTypeFactory` from the binding context.
+<2> Define the data type.
+<3> Define options.
```

Comment:
Optionally, we can add that available options depend on the data type. That we cannot have `normalizer` for a `double` type.